pygeoapi/provider/xarray.py

```diff
@@ -0,0 +1,435 @@
+        """
+
+        rangetype = {
+            'type': 'DataRecordType',
+            'field': []
+        }
+
+        for name, var in self._data.variables.items():
+            LOGGER.debug('Determing rangetype for {}'.format(name))
+
+            name, units = None, None
+            if len(var.shape) >= 3:
+                parameter = self._get_parameter_metadata(
+                    name, var.attrs)
+                name = parameter['description']
+                units = parameter['unit_label']
+
+                rangetype['field'].append({
+                    'id': name,
+                    'type': 'QuantityType',
+                    'name': var.attrs.get('long_name') or name,
+                    'definition': str(var.dtype),
+                    'nodata': 'null',
+                    'uom': {
+                        'id': 'http://www.opengis.net/def/uom/UCUM/{}'.format(
+                             units),
+                        'type': 'UnitReference',
+                        'code': units
+                    },
+                    '_meta': {
+                        'tags': var.attrs
+                    }
+                })
+
+        return rangetype
+
+    def query(self, range_type=[], subsets={}, format_='json'):
+        """
+         Extract data from collection collection
+
+        :param range_type: list of data variables to return (all if blank)
+        :param subsets: dict of subset names with lists of ranges
+        :param format_: data format of output
+
+        :returns: coverage data as dict of CoverageJSON or native format
+        """
+
+        if len(range_type) < 1:
+            range_type = self.fields
+
+        data = self._data[[*range_type]]
+
+        if(self._coverage_properties['x_axis_label'] in subsets or
+           self._coverage_properties['y_axis_label'] in subsets or
+           self._coverage_properties['time_axis_label'] in subsets):
+
+            LOGGER.debug('Creating spatio-temporal subset')
+
+            lon = self._coverage_properties['x_axis_label']
+            lat = self._coverage_properties['y_axis_label']
+            time = self._coverage_properties['time_axis_label']
+
+            query_params = {}
+            for key, val in subsets.items():
+                query_params[key] = slice(val[0], val[1])
+
+            data = data.sel(query_params)
+
+        out_meta = {'bbox': [
+            data.coords[self.lon_var].values[0],
+            data.coords[self.lat_var].values[0],
+            data.coords[self.lon_var].values[-1],
+            data.coords[self.lat_var].values[-1]],
+            "time": self._coverage_properties['time_range'],
+            "driver": "Xarray",
+            "height": data.dims[self.lat_var],
+            "width": data.dims[self.lon_var],
+            "time_steps": data.dims[self.time_var],
+            "variables": {var_name: var.attrs
+                          for var_name, var in data.variables.items()}
+        }
+
```

Comment:
We should ask functionality to return the coverage in the native format.  So if the config specifies:
```
              format:
                  name: GRIB2
                  mimetype: application/x-netcdf
```

See example in rasterio:
https://github.com/geopython/pygeoapi/blob/master/pygeoapi/provider/rasterio_.py#L254-L261

Reply:
I added the ability to download a native netCDF file.  Unfortunately it added a scipy dependency to write a memory file.